File: day-7/day7.pl
#!perl

use v5.24;
use warnings;
use List::Util ();

package Device::File {
	use Moo;
	has is_root => ( is => 'ro', default => !!0 );
	has name => ( is => 'ro', required => !!1 );
	has size => ( is => 'ro', required => !!1 );
	has parent => ( is => 'rw' );
	has total_size => ( is => 'lazy', builder => sub { shift->size } );
	sub full_path {
		my ( $self ) = @_;
		my $parent = $self->parent
			or return $self->name;
		return sprintf( '%s/%s', $parent->full_path, $self->name );
	}
	sub pretty_path {
		my ( $self ) = @_;
		return( $self->is_root ? '/' : $self->full_path );
	}
	sub display {
		my ( $self ) = @_;
		return sprintf(
			"%-84s %-4s %10d",
			$self->pretty_path,
			substr( ref($self), 8 ),
			$self->total_size,
		);
	}
	sub BUILD {
		my ( $self ) = @_;
		push( our @ALL, $self );
	}
}

package Device::Dir {
	use Moo;
	extends 'Device::File';
	has contents => ( is => 'ro', builder => sub { [] } );
	has '+size' => ( required => !!0, default => 0 );
	sub _build_total_size {
		my ( $self ) = @_;
		return List::Util::sum(
			$self->size,
			map( $_->total_size, $self->contents->@* ),
		);
	}
	sub add_child {
		my ( $self, @children ) = @_;
		$_->parent( $self ) for @children;
		push( $self->contents->@*, @children );
		return( wantarray ? @children : $children[0] );
	}
	sub make_child {
		my ( $self, $class, $name, %spec ) = @_;
		return $self->add_child( $class->new( name => $name, %spec ) );
	}
	sub get_child {
		my ( $self, $name ) = @_;
		my @found = grep( $_->name eq $name, $self->contents->@* );
		die if @found > 1;
		return $found[0];
	}
}

my $root = 'Device::Dir'->new( name => '', is_root => !!1 );
my $cwd  = $root;

my @lines = do { local @ARGV = 'input.txt'; <> };
chomp for @lines;

COMMAND_LINE: while ( @lines ) {
	my $command = shift @lines;

	if ( $command =~ m{^\$ cd /$} ) {
		$cwd = $root;
	}
	elsif ( $command =~ m{^\$ cd \.\.} ) {
		$cwd = $cwd->parent or die;
	}
	elsif ( $command =~ m{^\$ cd (.+)$} ) {
		my $name = $1;
		$cwd = $cwd->get_child( $name ) // $cwd->make_child( 'Device::Dir', $name );
	}
	elsif ( $command =~ m{^\$ ls$} ) {
		while ( @lines ) {
			if ( $lines[0] =~ m{^\$} ) {
				next COMMAND_LINE;
			}
			my $line = shift @lines;
			my ( $size, $name ) = split /\s+/, $line;
			next if $cwd->get_child( $name );
			if ( $size eq 'dir' ) {
				$cwd->make_child( 'Device::Dir', $name );
			}
			else {
				$cwd->make_child( 'Device::File', $name, size => $size );
			}
		}
	}
	else {
		die "Unknown input: '$command'!";
	}
}

say "Small dirs:";
my @small_dirs =
	grep $_->total_size <= 100_000,
	grep $_->isa( 'Device::Dir' ),
	@Device::File::ALL;
say $_->display
	for sort { $a->full_path cmp $b->full_path } @small_dirs;
say "";

my $total = List::Util::sum( map $_->total_size, @small_dirs );
say "Total of small dirs: $total";
say "";

say "=" x 100;
say "";

my $device_size = 70_000_000;
my $needed_space = 30_000_000;
my $current_space = $device_size - $root->total_size;
my $need_to_free = $needed_space - $current_space;
say "Current space on device: $current_space";
say "Needed space on device: $needed_space";
say "Need to free: $need_to_free";
say "";

my @candidates =
	sort { $a->total_size <=> $b->total_size }
	grep $_->total_size >= $need_to_free,
	@Device::File::ALL;
say "Candidates for deletion:";
say $_->display for @candidates;
say "";

say "Delete: ", $candidates[0]->pretty_path;

File: day-7/day7.php
<?php
// This is a direct port of day7.java.
// Honestly PHP is more elegant.

class DeviceFile {
	public $isRoot = false;
	public $name;
	public $size;
	public $parent;

	public function __construct( $n, $s ) {
		$this->isRoot = false;
		$this->parent = null;
		$this->name = $n;
		$this->size = $s;
	}

	public function type () {
		return "File";
	}

	public function totalSize () {
		return $this->size;
	}

	public function fullPath () {
		if ( $this->parent === null ) {
			return $this->name;
		}
		return $this->parent->fullPath() . "/" . $this->name;
	}

	public function prettyPath () {
		if ( $this->isRoot ) {
			return "/";
		}
		return $this->fullPath();
	}

	public function display () {
		return sprintf(
			"%-84s %-4s %10d",
			$this->prettyPath(),
			$this->type(),
			$this->totalSize()
		);
	}
}

class DeviceDir extends DeviceFile {
	public $contents;

	public function __construct( $n ) {
		parent::__construct( $n, 0 );
		$this->contents = [];
	}

	public function type () {
		return "Dir";
	}

	public function totalSize () {
		$total = $this->size;
		foreach ( $this->contents as $f ) {
			$total += $f->totalSize();
		}
		return $total;
	}

	public function makeChildDir ( $name ) {
		$child = new DeviceDir( $name );
		$child->parent = $this;
		$this->contents[] = $child;
		return $child;
	}

	public function makeChildFile ( $name, $size ) {
		$child = new DeviceFile( $name, $size );
		$child->parent = $this;
		$this->contents[] = $child;
		return $child;
	}

	public function getChild ( $name ) {
		foreach ( $this->contents as $f ) {
			if ( $f->name == $name ) {
				return $f;
			}
		}
		return null;
	}

	public function allDirs () {
		$all = [ $this ];
		foreach ( $this->contents as $d ) {
			if ( $d instanceof DeviceDir ) {
				foreach ( $d->allDirs() as $child ) {
					$all[] = $child;
				}
			}
		}
		return $all;
	}
}

class DeviceFS {

	public $root;

	public function __construct () {
		$root = new DeviceDir( "" );
		$root->isRoot = true;
		$this->root = $root;
	}

	public static function fromScript ( $infile ) {

		$me = new DeviceFS();
		$cwd = $me->root;

		$lines = file( $infile );

		$cd_pattern = "/^\\$ cd (.+)$/";

		while ( count( $lines ) ) {
			$line = trim( array_shift( $lines ) );

			if ( $line == '$ cd /' ) {
				$cwd = $me->root;
			}
			elseif ( $line == '$ cd ..' ) {
				$cwd = $cwd->parent;
			}
			elseif ( $line == '$ ls' ) {
				while ( count( $lines ) ) {
					if ( substr( $lines[0], 0, 1 ) == '$' ) {
						continue 2;
					}
					list ( $size, $name ) = explode( " ", trim( array_shift( $lines ) ) );
					if ( null === $cwd->getChild( $name ) ) {
						if ( $size == 'dir' ) {
							// cwd.makeChildDir( name );
						}
						else {
							$cwd->makeChildFile( $name, $size );
						}
					}
				}
			}
			elseif ( preg_match( $cd_pattern, $line, $matches ) ) {
				$name = $matches[1];
				$child = $cwd->getChild( $name );
				if ( $child === null ) {
					$child = $cwd->makeChildDir( $name );
				}
				$cwd = $child;
			}
			else {
				echo "Unhandled line: '" . $line . "'!\n";
			}
		}

		return $me;
	}
}

$fs = DeviceFS::fromScript( "input.txt" );
$all_dirs = $fs->root->allDirs();

$small_dirs = array_filter( $all_dirs, function ( $d ) {
	return $d->totalSize() <= 100000;
} );
usort( $small_dirs, function ( $d1, $d2 ) {
	return $d1->totalSize() <=> $d2->totalSize();
} );

$small_total = 0;
echo "Small dirs:\n";
foreach ( $small_dirs as $d ) {
	echo $d->display() . "\n";
	$small_total += $d->totalSize();
}
echo "Total of small dirs: $small_total\n";
echo "\n";

echo "================\n";
echo "\n";

$device_size = 70_000_000;
$needed_space = 30_000_000;
$current_space = $device_size - $fs->root->totalSize();
$need_to_free = $needed_space - $current_space;

echo "Current space on device: $current_space\n";
echo "Needed space on device: $needed_space\n";
echo "Need to free: $need_to_free\n";
echo "\n";

$big_dirs = array_filter( $all_dirs, function ( $d ) use ( $need_to_free )  {
	return $d->totalSize() >= $need_to_free;
} );
usort( $big_dirs, function ( $d1, $d2 ) {
	return $d1->totalSize() <=> $d2->totalSize();
} );

echo "Candidates for deletion:\n";
foreach ( $big_dirs as $d ) {
	echo $d->display() . "\n";
}
echo "\n";
echo "Delete: " . $big_dirs[0]->prettyPath() . "\n";
